kernel/src/process/program_loader/elf/load_elf.rs

```diff
@@ -225,66 +217,172 @@ fn map_segment_vmos(
         aligned_range.start + start_in_page_offset..aligned_range.end - end_in_page_offset
     } else {
         // Not relocatable object. Map as-is.
+        elf_va_range.clone()
+    };
+
+    let relocated_range =
+        RelocatedRange::new(elf_va_range, map_range.start).expect("Mapped range overflows");
+
+    let mut start_data = !0;
+    let mut end_data = 0;
```

Comment:
I propose to add a new method to `ElfHeaders` as follows:
```rust
    /// Calculates the virtual address bounds of data segments as a range.
    ///
    /// Note that this method only considers the data portion. In other words, it accounts for the
    /// file size of the segment instead of the memory size.
    pub(super) fn calc_data_vaddr_bounds(&self) -> Option<Range<Vaddr>> {
```

Check `ElfHeaders::calc_total_vaddr_bounds` for an example. I think it can be clearer than this form.